### ISSUE TITLE
Couldn't determine Mocha version

### ISSUE THREAD

**Issue by fr0** [cypress-io/cypress#4602]
### Current behavior:

```
Couldn't determine Mocha version
```

This has been printing at the beginning of my test since the upgrade to 3.3.2. This did not happen with Cypress 3.3.1. I don't have `mocha` or `@types/mocha` in my `package.json` file's devDependencies; should I?

### Versions

Cypress 3.3.2

**Comment by bahmutov** [comment]
hmm, that does seem weird, we don't see this message in any of our test repositories. Any details about the tests that might give us a clue?

**Comment by fr0** [comment]
I can reproduce it here:  https://github.com/fr0/cypress-slowdown

**Comment by bahmutov** [comment]
yup, I can see it

<img width="1128" alt="Screen Shot 2019-06-28 at 10 52 38 AM" src="https://user-images.githubusercontent.com/2212006/60351051-ea7f5880-9992-11e9-8a21-067be65c7f87.png">

**Comment by bahmutov** [comment]
The message goes away when I remove from `cypress.json` option "reporter: junit" or if I use any other built-in reporter

<img width="1138" alt="Screen Shot 2019-06-28 at 10 57 33 AM" src="https://user-images.githubusercontent.com/2212006/60351346-945ee500-9993-11e9-8ec5-25efd78315de.png">

**Comment by bahmutov** [comment]
in v3.3.2 we did upgrade mocha-junit-reporter from 1.18.0 to 1.23.0 for #4417 see https://docs.cypress.io/guides/references/changelog.html#3-3-2

**Comment by bahmutov** [comment]
looking at `mocha-junit-reporter` relevant code bundled with Cypress
```
var createStatsCollector;
var mocha6plus;

try {
  var json = JSON.parse(
    fs.readFileSync("./node_modules/mocha/package.json", "utf8")
  );
  version = json.version;
  if (version >= "6") {
    createStatsCollector = require("mocha/lib/stats-collector");
    mocha6plus = true;
  } else {
    mocha6plus = false;
  }
} catch (e) {
  console.warn("Couldn't determine Mocha version");
}
module.exports = MochaJUnitReporter;
```

**Comment by bahmutov** [comment]
printing the error, yeah it cannot load `mocha` like this!
```
$ npx cypress run
current cwd /Users/gleb/git/cypress-slowdown
Couldn't determine Mocha version
{ Error: ENOENT: no such file or directory, open './node_modules/mocha/package.json'
    at Object.fs.openSync (fs.js:646:18)
    at Object.module.(anonymous function) [as openSync] (ELECTRON_ASAR.js:166:20)
    at Object.fs.readFileSync (fs.js:551:33)
    at Object.fs.readFileSync (ELECTRON_ASAR.js:538:29)
    at Object.<anonymous> (/Users/gleb/Library/Caches/Cypress/3.3.2/Cypress.app/Contents/Resources/app/packages/server/node_modules/mocha-junit-reporter/index.js:18:8)
    at Object.<anonymous> (/Users/gleb/Library/Caches/Cypress/3.3.2/Cypress.app/Contents/Resources/app/packages/server/node_modules/mocha-junit-reporter/index.js:455:3)
```

The breaking commit that introduced this check https://github.com/michaelleeallen/mocha-junit-reporter/commit/b28d669995af194968dcbf5301ff681797b8a26e

I wish we could just pass the mocha version instead of this file system check

**Comment by bahmutov** [comment]
we will need to think how to solve this, probably will need to fork the junit reporter module ...

**Comment by jennifer-shehane** [comment]
There's an issue open for this here. https://github.com/michaelleeallen/mocha-junit-reporter/issues/87

**Comment by YOU54F** [comment]
FWIW, if it can’t read the version, it prints the warning but reverts back to pre mocha6 behaviour, so the reporter should behave as normal with Cypress as it ships with quite an old version of mocha iirc, so bar the warning, there should be no difference in functionality.

I’ll have a think to a solution for the use case in issue 87 on mocha junit reporter but it’s a difficult one, for users of mocha6+ when running a cypress project in a different folder. Maybe an option needs to be passed into mocha-junit-reporter that tells it where the project is stored

**Comment by gcirone** [comment]
@YOU54F well this is a pain because the process fail at the end 

```
TypeError: Cannot read property 'failures' of undefined
    at MochaJUnitReporter.getXml (/Users/gcirone/Works/sports/dolores-sports/node_modules/mocha-junit-reporter/index.js:421:25)
    at MochaJUnitReporter.flush (/Users/gcirone/Works/sports/dolores-sports/node_modules/mocha-junit-reporter/index.js:350:18)
    at MochaJUnitReporter.<anonymous> (/Users/gcirone/Works/sports/dolores-sports/node_modules/mocha-junit-reporter/index.js:227:10)
    at emitOne (events.js:121:20)
    at Runner.emit (events.js:211:7)
    at Reporter.emit (/Users/gcirone/Library/Caches/Cypress/3.3.1/Cypress.app/Contents/Resources/app/packages/server/lib/reporter.js:239:55)
    at Object.server.startWebsockets.onMocha (/Users/gcirone/Library/Caches/Cypress/3.3.1/Cypress.app/Contents/Resources/app/packages/server/lib/project.js:298:22)
    at Socket.<anonymous> (/Users/gcirone/Library/Caches/Cypress/3.3.1/Cypress.app/Contents/Resources/app/packages/server/lib/socket.js:235:36)
    at emitTwo (events.js:126:13)
    at Socket.emit (events.js:214:7)
    at /Users/gcirone/Library/Caches/Cypress/3.3.1/Cypress.app/Contents/Resources/app/packages/socket/node_modules/socket.io/lib/socket.js:503:12
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```

Anyway IMHO should be another way to detect the mocha version and should be transparent for the user, isn't user responsibility set the option to know where mocha is located.

With the __dirname workaround is working perfect for both case: cypress.json and package.json in root and in a sub folders.

This is the simple workaround on `postinstall` that i have
`sed -i 's/".\/node_modules/__dirname + "\/../g' $MOCHA_JUNIT_FILE`

**Comment by gcirone** [comment]
I just suggest a possible solution that work for all cases.

```
const mochaPackageJson = path.dirname(require.resolve('mocha')) + '/package.json';
const { version } = require(mochaPackageJson);
console.log('version', version)
```

https://github.com/michaelleeallen/mocha-junit-reporter/issues/87#issuecomment-508037669

**Comment by gcirone** [comment]
PR opened https://github.com/michaelleeallen/mocha-junit-reporter/pull/88 👍

**Comment by gcirone** [comment]
Released in mocha-junit-reporter@1.23.1 work as expected 🥇

**Comment by jennifer-shehane** [comment]
Released in [3.4.1](https://on.cypress.io/changelog#3-4-1)
